### PR TITLE
Xtrabackup 8

### LINKED ISSUE #263: support percona xtrabackup 8.0
Submitting on behalf of MZ

Xtrabackup 8.0 has been introduced to be compatible with MySQL 8.0. This version has removed the innobackupex script and now uses the xtrabackup binary exclusively for all functions. The xtrabackup binary requires different options to achieve the same result as innobackupex. Namely, to execute a backup, you need to always start the command with:

/bin/xtrabackup --backup --target=/path/to/datadir [additional options]

Furthermore, you no longer need to append the target path at the end of the command. This means that for Holland-xtrabackup, altering the following in the backupset conf will result in an invalid backup command:

innobackupex = xtrabackup

# holland bk xtrabackup
[...]
Executing: /bin/xtrabackup --defaults-file=/data1/holland_backup/xtrabackup/20190131_195352/my.cnf --stream=xbstream --tmpdir=/data1/holland_backup/xtrabackup/20190131_195352 --no-lock --no-timestamp --parallel=4 /data1/holland_backup/xtrabackup/20190131_195352
  > /data1/holland_backup/xtrabackup/20190131_195352/backup.xb.zst 2 > /data1/holland_backup/xtrabackup/20190131_195352/xtrabackup.log
[...]
    ! xtrabackup: Error: unknown argument: '/data1/holland_backup/xtrabackup/20190131_195352'

A properly arranged command should be along the lines of:

/bin/xtrabackup --backup --target=/data1/mysql8/mysql --stream=xbstream --no-lock --no-timestamp --parallel=4 2> /data1/holland_backup/xtrabackup/20190131_195352/xtrabackup.log | zstd -T8 > /data1/holland_backup/xtrabackup/20190131_195352/backup.xb.zst

### PR DESCRIPTION
**Pull request by soulen3**
Fixes #263

### CODE REVIEW

**Comment by soulen3** [comment]
I've updated the xtrabackup module to detect if the 'xtrabackup' binary is a version greater than 8.0.0, and it changes the argument set based on that. This should allow the default configuration for both multiple versions of xtrabackup.

You can see the relevant code here:
https://github.com/holland-backup/holland/blob/xtrabackup_8/plugins/holland.backup.xtrabackup/holland/backup/xtrabackup/util.py?#L210-#L248

```
# holland -d  bk xtrabackup
Adding plugin directory: u'/usr/share/holland/plugins'
Holland 1.1.11 started with pid 10621
--- Starting backup run ---
Set advisory lock on /etc/holland/backupsets/xtrabackup.conf
Creating backup path /var/spool/holland/xtrabackup/20190204_173249
* Generating mysql option file: /var/spool/holland/xtrabackup/20190204_173249/my.cnf
  + Added !include /etc/my.cnf
  + Added !include /root/.my.cnf
Writing out config to /var/spool/holland/xtrabackup/20190204_173249/backup.conf
Debug: Database backups writen to /var/lib/mysql//performance_schema
Debug: Database backups writen to /var/lib/mysql//#innodb_temp
Debug: Database backups writen to /var/lib/mysql//mysql
Debug: Database backups writen to /var/lib/mysql//sys
Estimated Backup Size: 166.12MB
Using Historic Space Estimate: Checking for information in /var/spool/holland/xtrabackup/newest/backup.conf
Last backup used 2.30MB
The old and new backup estimate are roughly the same size, use old backup size for new size estimate
Adjusting estimated size to last backup total * 1.1: 2.53MB
Starting backup[xtrabackup/20190204_173249] via plugin xtrabackup
Executing: /bin/xtrabackup --version
xtrabackup: recognized server arguments: --datadir=/var/lib/mysql
/bin/xtrabackup version 8.0.4 based on MySQL server 8.0.13 Linux (x86_64) (revision id: c2c0777)
Use xtrabackup without innobackupex
* Executing: /bin/gzip -1
Executing: /bin/xtrabackup --defaults-file=/var/spool/holland/xtrabackup/20190204_173249/my.cnf --backup --stream=xbstream --no-timestamp
  > /var/spool/holland/xtrabackup/20190204_173249/backup.tar.gz 2 > /var/spool/holland/xtrabackup/20190204_173249/xtrabackup.log
Skipping --prepare/--apply-logs since backup is streamed
Final on-disk backup size 2.30MB
1.38% of estimated size 166.12MB
Writing out config to /var/spool/holland/xtrabackup/20190204_173249/backup.conf
Backup completed in 3.25 seconds
Purged xtrabackup/20190204_173241
1 backups purged
Released lock /etc/holland/backupsets/xtrabackup.conf
--- Ending backup run ---
```

File: plugins/holland.backup.mariabackup/holland/backup/mariabackup/util.py [hunk 1]
```diff
@@ -116,7 +116,7 @@ def apply_mariabackup_logfile(mb_cfg, backupdir):
     innobackupex = mb_cfg["innobackupex"]
     if not isabs(innobackupex):
         innobackupex = which(innobackupex)
-    args = [innobackupex, "--prepare", backupdir]
+    args = [innobackupex, "--prepare", "--target-dir=" + backupdir + "/data"]
```
Comment (nwbreneman) [inline plugins/holland.backup.mariabackup/holland/backup/mariabackup/util.py]: We should probably use `os.path.join` here just to be safe; e.g `args = [innobackupex, "--prepare", "--target-dir=" + path.join(backupdir, "/data")]`

File: plugins/holland.backup.xtrabackup/holland/backup/xtrabackup/util.py [hunk 2]
```diff
@@ -95,10 +98,14 @@ def apply_xtrabackup_logfile(xb_cfg, backupdir):
         LOG.warning("Skipping --apply-logs since --compress option appears " "to have been used.")
         return
 
-    innobackupex = xb_cfg["innobackupex"]
-    if not isabs(innobackupex):
-        innobackupex = which(innobackupex)
-    args = [innobackupex, "--apply-log", backupdir]
+    if binary_xtrabackup:
+        innobackupex = which("xtrabackup")
+        args = [innobackupex, "--prepare", "--target-dir=" + backupdir + "/data"]
```
Comment (nwbreneman) [inline plugins/holland.backup.xtrabackup/holland/backup/xtrabackup/util.py]: Another case for `os.path.join` to be used.

File: plugins/holland.backup.xtrabackup/holland/backup/xtrabackup/util.py [hunk 3]
```diff
@@ -95,10 +98,14 @@ def apply_xtrabackup_logfile(xb_cfg, backupdir):
...
+    else:
+        innobackupex = xb_cfg["innobackupex"]
+        if not isabs(innobackupex):
+            innobackupex = which(innobackupex)
+        args = [innobackupex, "--apply-log", backupdir + "/data"]
```
Comment (nwbreneman) [inline plugins/holland.backup.xtrabackup/holland/backup/xtrabackup/util.py]: Another case for `os.path.join` to be used.

**Pull request merged**